### PR TITLE
feat(mailbox): adding user api

### PR DESCRIPTION
**Pull request by andrewxhill**

### CODE REVIEW

File: packages/users/src/users.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  async getThread(name: string): Promise<GetThreadReply.AsObject> {
+    return getThread(this, name)
+  }
+
+  async setupMailbox(): Promise<{ mailboxID: Uint8Array | string }> {
+    return setupMailbox(this)
+  }
+
+  async sendMessage(
```
Comment (andrewxhill) [inline packages/users/src/users.ts:93]: This is intended to match the API found here, https://github.com/textileio/textile/blob/5d8532b244aa0628ce401c11cf42c8e0953d6aeb/api/users/client/client.go#L128 but the libp2p crypto types don't have all the necessary functions to fulfill this.

File: packages/users/src/api/index.ts [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+  const req = new DeleteMessageRequest()
+  req.setId(id)
+  await api.unary(API.DeleteSentboxMessage, req, ctx)
+  return
+}
+
+/**
+ * @internal
+ */
+export function watchMailbox(
```
Comment (andrewxhill) [inline packages/users/src/api/index.ts:287]: new listen method for mailbox @carsonfarmer

File: packages/users/src/users.spec.ts [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+      rec = await user1.listSentboxMessages()
+      expect(rec.length).to.equal(0)
+    })
+    it('mailboxID should exist', async () => {
+      const user1 = new Users(user1Ctx)
+      const mailboxID = await user1.getMailboxID()
+      expect(mailboxID).to.not.be.undefined
+    })
+
+    it('inbox listen should return new messages', (done) => {
```
Comment (andrewxhill) [inline packages/users/src/users.spec.ts:313]: new listen tests, this one 'til the end @carsonfarmer

File: etc/hub.api.md [hunk 4]
```diff
@@ -48,10 +64,12 @@ export { ArchiveStatusReply }
 
 export { ArchiveWatchReply }
 
+// Warning: (ae-incompatible-release-tags) The symbol "Buckets" is marked as @public, but its signature references "GrpcAuthentication" which is marked as @internal
```
Comment (carsonfarmer) [inline etc/hub.api.md:67]: Is this an issue?

File: packages/buckets/src/api/index.ts [hunk 5]
```diff
@@ -55,51 +55,84 @@ export interface PushPathResult {
...
-   * @param context The context to use for interacting with the APIs. Can be modified later.
-   */
-  constructor(public context: ContextInterface = new Context(), debug = false) {
-    this.serviceHost = context.host
-    this.rpcOptions = {
-      transport: WebsocketTransport(),
-      debug,
-    }
+/**
+ * Response from bucket links query.
```
Comment (carsonfarmer) [inline packages/buckets/src/api/index.ts:59]: I know is a pain to maintain these outside the gRPC types, but wow does this improve the DX considerably. Too bad there isn't a way to derive these automatically from the gRPC responses?

File: packages/buckets/src/buckets.ts [hunk 6]
```diff
@@ -103,34 +99,73 @@ const logger = log.getLogger('buckets')
  * }
  * ```
  */
-export class Buckets extends BucketsGrpcClient {
+export class Buckets extends GrpcAuthentication {
   /**
-   * Creates a new gRPC client instance for accessing the Textile Buckets API.
-   * @param auth The user auth object.
+   * {@inheritDoc @textile/hub#GrpcAuthentication.copyAuth}
```
Comment (carsonfarmer) [inline packages/buckets/src/buckets.ts:104]: Cool!

File: packages/crypto/src/keypair.ts [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+  return new TextDecoder().decode(encoded)
+}
+
+/**
+ * Default implementation of the {@link Public} interface, with encryption
+ * support.
+ * In theory, RSA, ed25519, and secp256k1 key types (and more) should be
+ * supported, although currently only ed25519 has full verify and encrypt
+ * capabilities.
+ * @privateRemarks Support additional key types by default. For now, sticking to ed25519
```
Comment (carsonfarmer) [inline packages/crypto/src/keypair.ts:23]: Cool!

File: packages/crypto/src/keypair.ts [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+ * import { PrivateKey } from '@textile/hub'
+ *
+ * function example() {
+ *   const identity = PrivateKey.fromRandom()
+ *   const publicKey = identity.public.toString()
+ *   return publicKey
+ * }
+ * ```
+ *
+ * @example
```
Comment (carsonfarmer) [inline packages/crypto/src/keypair.ts:41]: Love these!

File: packages/users/src/api/index.ts (hunk 2 above)
Comment (carsonfarmer) [reply packages/users/src/api/index.ts:287]: Sweet, yup looks good. One day we'll get those Buffer.from's out of everything.

File: etc/hub.api.md (hunk 4 above)
Comment (andrewxhill) [reply etc/hub.api.md:67]: I hid it intentionally so it doesn't make the docs unclear. This warning just means that where docs reference it now it wont be linked to anything. So in particular, when it says Users class is an instance of this, it's just plaintext

**Pull request merged**